### PR TITLE
completed query function to find available meeting times

### PR DESCRIPTION
**Pull request by hjohns28**
parent pull request: #32 (because accidentally did this work while still on the last branch, efficiencychanges branch, and then took off that commit and moved it to this branch)

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,223 @@
...
+      return allAttendeeAvailability;
+    } 
+
+    //if the meeting is scheduled for longer than a day, it cannot be scheduled
+    else if (request.getDuration() > 24*60) {
+      return allAttendeeAvailability;
+    } 
+
+    //if there are no meetings, any time of day works 
+    else if (events.size() == 0) {
```
Comment (kristen-law) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This case can be considered in the first `if` statement also to avoid duplicate code (they both return the whole day time range).

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,223 @@
...
+import java.util.*; 
+import java.util.Comparator;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    ArrayList<TimeRange> mandatoryAttendeeAvailability = new ArrayList<>();
+    ArrayList<TimeRange> allAttendeeAvailability = new ArrayList<>();
```
Comment (kristen-law) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This variable type should generally be declared as `List` instead of `ArrayList`. `ArrayList` is a specific implementation of the `List` interface. It's generally preferred to store the variable as a `List` so you're not locked into a specific implementation. (Just adding the comment here, but please fix the others also.)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,223 @@
...
+        } else {
+          i++;
+        }
+      }
+
+      //copy only the time ranges from the meetings, and sort 
+      ArrayList<TimeRange> mandatoryAttendeeMeetingTimes = new ArrayList<>();
+      for (int b = 0; b < mandatoryAttendeeMeetings.size(); b++) {
+        mandatoryAttendeeMeetingTimes.add(mandatoryAttendeeMeetings.get(b).getWhen());
+      }
```
Comment (kristen-law) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This part seems a little overly complex. Instead of initializing `mandatoryAttendeeMeetings` with all the events, removing the events they're not a part of, and then copying each meeting's time range to another `mandatoryAttendeeMeetingTimes` array, you could just loop through the events, check if the mandatory attendees are a part of the event, and if so, then add that meeting time to `mandatoryAttendeeMeetingTimes`.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,223 @@
...
+      //remove any time ranges that aren't long enough 
+      int m = 0;
+      while (m < mandatoryAttendeeAvailability.size()) {
+        if (mandatoryAttendeeAvailability.get(m).duration() < request.getDuration()) {
+          mandatoryAttendeeAvailability.remove(m);
+          continue;
+        } else {
+          m++;
+        }
+      }
```
Comment (kristen-law) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Instead of looping through `mandatoryAttendeeAvailability` here to remove time ranges that aren't long enough, you could check before you even add it to `mandatoryAttendeeAvailability` (like where you check that the duration > 0 above).

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,223 @@
...
+
+      //if there are optional attendees, find the availability for all attendees, and compare to 
+      //the availability for mandatory attendees 
+      if (request.getOptionalAttendees().size() > 0) {
```
Comment (kristen-law) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It looks like there's a lot of code in this section that just duplicates what's been done above for mandatory attendees. To avoid duplicating code, I'd suggest creating helper functions that you can use for both cases.

**Review by kristen-law — CHANGES_REQUESTED** [review]
Still reviewing, just wanted to send over some comments first.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I agree, I was thinking that I was duplicating too much code. Thanks for the feedback, I'll create helper functions now!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:28]: Sorry but just to clarify, for "this variable type" do you mean only the ArrayLists of TimeRanges, or does this revision apply to all of the ArrayLists that I created in this function? I'm guessing the latter, but just wanted to check. Also, thank you for the information, that's good to know!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Good point, this is definitely overly complex. I just have a quick question, if that's okay. The reason I copied the meetings from the collection events to an ArrayList was that I couldn't figure out a way to access the different attributes, like getAttendees() or getWhen() directly from the collection. I could iterate through it, using an Iterator, but I couldn't access the values (I looked up ways to access objects in a collection but didn't find anything that solved this problem, only how to access the entire object). Would it be okay to first copy the meetings to an ArrayList, and then simplify it like you said by looping through that ArrayList and directly adding the times to mandatoryAttendeeMeetingTimes if there are overlapping attendees?

Comment (kristen-law) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: If you use an Iterator to iterate through `events`, you'd be able to get the next `Event` in the collection with `iterator.next()`. If you assign that to a variable (e.g. `Event event = iterator.next();`), you should be able to call getAttendees() or getWhen() on `event`. Alternatively, you could use a for each-loop: `for (Event event : events)`.

Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Okay, thank you so much. I was trying to do it all in one step (interator.next().getAttendees()) instead of saving it as a variable first. Sorry I didn't figure that out sooner on my own!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,135 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.*;
```
Comment (FabriceCastel) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Please import only the specific utils you need (as opposed to all of them)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,135 @@
...
+import java.util.Set;
+import java.util.*; 
+import java.util.Comparator;
+import java.util.Collections;
+import java.util.Iterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
```
Comment (FabriceCastel) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Refer to bullet 2 from [this whitespace usage section of the style guide](https://google.github.io/styleguide/jsguide.html#formatting-whitespace)

```
A single blank line appears:

1. Between consecutive methods in a class or object literal
  a. Exception: A blank line between two consecutive properties definitions in an object literal (with no other code between them) is optional. Such blank lines are used as needed to create logical groupings of fields.

2. Within method bodies, sparingly to create logical groupings of statements. Blank lines at the start or end of a function body are not allowed.

3. Optionally before the first or after the last method in a class or object literal (neither encouraged nor discouraged).

4. As required by other sections of this document (e.g. 3.6 goog.require and goog.requireType statements).
```

TL;DR no newlines at the start and end of function definitions

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -15,9 +15,135 @@
...
+    List<TimeRange> mandatoryAttendeeAvailability = new ArrayList<>();
+    List<TimeRange> allAttendeeAvailability = new ArrayList<>();
+
+    //if the meeting is scheduled for longer than a day, it cannot be scheduled
+    if (request.getDuration() > 24*60) {
+      return allAttendeeAvailability;
+    } 
+
+    //if no attendees or no meetings, any time of day works
+    else if ((request.getAttendees().size() == 0 && request.getOptionalAttendees().size() == 0) || events.size() == 0) {
```
Comment (FabriceCastel) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I haven't found a solid reference for this in particular in the style guide but normally you'd write any `else` or `else if` clause on the same line as the previous closing bracket, and move any comments inside, eg.

```
if (foo) {
  ...
} else if (bar) {
  // the user has bar-ed
  ...
}
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -15,9 +15,135 @@
...
+      allAttendeeAvailability.add(TimeRange.WHOLE_DAY);
+      return allAttendeeAvailability;
+    } 
+
+    //otherwise, find all possible meeting times   
+    else {
```
Comment (FabriceCastel) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Same comment about the closing bracket + `else`

You'd also usually not use a newline at the start or end of a code block (much like for function definitions), ie.

```
if (foo) {
   code.starts("here");
}
```

as opposed to

```
if (bar) {

  not.likeThis();
}
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -15,9 +15,135 @@
...
+        conflictingMeetings.add(event.getWhen());
+      }
+    }
+    
+    Collections.sort(conflictingMeetings, Comparator.comparingInt(TimeRange::start)); 
+    return conflictingMeetings;
+  }
+
+  //remove nested and merge overlapping time ranges
+  public void fixNestedOrOverlappingMeetings(List<TimeRange> meetingTimes) {
```
Comment (FabriceCastel) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This might be more succinctly named `mergeOverlappingMeetings`, since that's exactly what it does. It makes it a bit less ambiguous since "fixing" those meetings could mean different things.

**Review by FabriceCastel — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java [hunk 11]
```diff
@@ -46,6 +46,9 @@ public MeetingRequest(Collection<String> attendees, long duration) {
    * Returns a read-only copy of the people who are optional to attend this meeting.
    */
   public Collection<String> getOptionalAttendees() {
+    if (optional_attendees == null) {
+      return Collections.unmodifiableCollection(new HashSet<>());
+    }
```
Comment (kristen-law) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: Why was this added here? `optional_attendees` should never be null. It's initialized to a new HashSet at the top of the class.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks, I was trying to think of a better name because I new it was a little verbose and unclear, but hadn't changed it. And as for all of my styling errors, I was planning on going back to check the guide earlier, but then had a meeting. I should have not submitted the pull request in the first place until I had checked on the styling, so I apologize, I just meant to say that I recognized that I messed up some things. I appreciate the feedback!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -15,9 +15,135 @@
...
+      } 
+      else if (currentMeetingTime.overlaps(nextMeetingTime)) {
+        TimeRange totalMeetingTime = TimeRange.fromStartEnd(currentMeetingTime.start(), nextMeetingTime.end(), false);
+        meetingTimes.set(j, totalMeetingTime);
+        meetingTimes.remove(j+1);
+        continue;
+      } 
+      else {
+        j++;
+      }
```
Comment (kristen-law) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You don't need to add `continue` on lines 103 and 109. This isn't like a switch case where execution will fall through: after the condition for one if-else case is met, it will execute the code for that case and then jump to the end of the if-else block. In this case, that's also at the end of the while loop so it will return to the top of the loop.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -15,9 +15,135 @@
...
+    List<TimeRange> conflictingMeetings = new ArrayList<>();
+    
+    for (Iterator iterator = events.iterator(); iterator.hasNext();) {
+      List<String> newMeetingAttendees = new ArrayList<>();
+      if (allAttendees) {
+        newMeetingAttendees.addAll(0, request.getOptionalAttendees());
+      }
+      newMeetingAttendees.addAll(0, request.getAttendees());
+
+      Event event = (Event) iterator.next();
```
Comment (kristen-law) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: As you probably figured out here, the reason you couldn't do` iterator.next().getAttendees()` previously is because `iterator.next()` was returning an Object instead of an Event, and the Object did not have the `getAttendees()` method. Casting that Object to an Event allows you to access that method.

However, when you define the iterator on line 74, you could write `Iterator<Event> iterator` to specify the iterator's element type as an Event. This would mean the return type of `iterator.next()` would be an Event without any casting. In that case, you would be able to use `iterator.next().getAttendees()` without running into any errors. (Since you need to access the event twice though for `getAttendees()` and `getWhen()`, it's best to store `iterator.next()` in the `event` variable because you don't want to call it twice.)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -15,9 +15,135 @@
...
+        if (allAttendeeAvailability.size() > 0 || request.getAttendees().size() == 0) {
+          return allAttendeeAvailability;
+        }
+      }
+    }
+    return mandatoryAttendeeAvailability;
+  }
+
+  //only keep meetings that have overlapping attendees, and sort by start time 
+  public List<TimeRange> findAndSortConflictingMeetings(Collection<Event> events, MeetingRequest request, Boolean allAttendees) {
```
Comment (kristen-law) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Since these helper functions are only used inside this class, it's best to limit their visibility to private instead of making it public.

**Review by kristen-law — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java (hunk 11 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: Oh yeah, we were only supposed to make changes to the query function, but when I tried to run the program on the local server I kept getting an error. I asked in the technical chat and others all had the same error, and they said to fix it by adding this (checking if null). I can review this and change it - I had just wanted to make sure that the function I was supposed to fill out was working properly.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: whoops, I don't know why I didn't catch that. I don't think I had it as if / else if before. My bad, thank you!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 13 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I see, thank you for all of the information - I really appreciate it and that's good to know! That makes sense.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java (hunk 11 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: Okay, I think the error was because the constructor was being called, but optional_attendees isn't in the constructor. I changed it so that it is initialized to a new HashSet in the constructor and there's no error. Hopefully this is okay?

Comment (kristen-law) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: Interesting, I don't see how that would have fixed the issue. Initializing the instance variable at declaration vs. in the constructor should have the same result.

Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: Yeah, I'm not entirely clear on this either, it seems like initializing it above should work fine. I'm wondering if it is because in the javascript, it calls the constructor directly with the parameters for attendees and duration, so it only sets the values for attendees and duration?

Comment (kristen-law) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: Calling the constructor creates the `MeetingRequest` object which will have its instance variables `attendees` and `optional_attendees` initialized to an empty HashSet at that time. If they weren't initialized, you wouldn't be able to call `this.attendees.addAll(attendees);` in the constructor because `attendees` would be null. I'm guessing you're not running into an error because you still have the `null` check in getOptionalAttendees (although it's still not clear why that check is necessary).

Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: oh man, I meant to remove that part. That makes perfect sense what you are saying. Sorry for my mess up and re-requesting review too soon. I will continue to see if I can understand what the problem is!

Comment (FabriceCastel) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: @hjohns28 out of curiosity, could you copy/paste the full error message you got before putting in this fix?

**Review by FabriceCastel — APPROVED** [review]
LGTM

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java (hunk 11 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: of course! 

script.js:57 POST https://8080-37040b47-fb20-4338-b06b-f8cf37e1cecd.us-west1.cloudshell.dev/query 500 (Internal Server Error)
queryServer @ script.js:57
sendMeetingRequest @ script.js:31
onclick @ ?authuser=0:28
8080-37040b47-fb20-4338-b06b-f8cf37e1cecd.us-west1.cloudshell.dev/:1 Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 0
Promise.then (async)
sendMeetingRequest @ script.js:31
onclick @ ?authuser=0:28

Comment (kristen-law) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: Thanks for the error message. I'll LGTM this PR because the issue isn't within your own code. You can leave the null check in so that you don't get that error.

I'm guessing the error is the following:
The `MeetingRequest` class defined in script.js doesn't have the `optional_attendees` property defined (it only has `duration` and `attendees`). When that information is sent to the `QueryServlet` it is converted to a `MeetingRequest` defined by MeetingRequest.java using `gson.fromJson` (not the constructor). Since the Javascript `MeetingRequest` didn't have `optional_attendees` initialized, when it is converted to the Java `MeetingRequest` class, `optional_attendees` is null.

Happy to chat about this further if that's not clear.

**Review by kristen-law — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java (hunk 11 above)
Comment (hjohns28) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:52]: That makes sense - I was wondering about the MeetingRequest class in the script.js file. I might look at the project a little more, but I understand what you're saying and think that's likely. Thank you for the explanation and help, I appreciate it!
